src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java

```diff
@@ -1577,31 +1589,35 @@ protected boolean isDependencyChanged() {
     }
 
     /**
-     * @param classPathEntry entry to check
+     * @param file entry to check
      * @param buildStartTime time build start
      * @return if any changes occurred
      */
-    private boolean hasNewFile(File classPathEntry, Date buildStartTime) {
-        if (!classPathEntry.exists()) {
-            return false;
-        }
-
-        if (classPathEntry.isFile()) {
-            return classPathEntry.lastModified() >= buildStartTime.getTime()
-                    && fileExtensions.contains(FileUtils.getExtension(classPathEntry.getName()));
-        }
-
-        File[] children = classPathEntry.listFiles();
-
-        for (File child : children) {
-            if (hasNewFile(child, buildStartTime)) {
-                return true;
+    private boolean hasNewFile(Path file, Instant buildStartTime) {
+        if (Files.isRegularFile(file) && fileExtensions.contains(getFileExtension(file))) {
+            try {
+                Instant lastModifiedTime =
+                        Files.getLastModifiedTime(file).toInstant().truncatedTo(ChronoUnit.MILLIS);
+                boolean isChanged = lastModifiedTime.isAfter(buildStartTime);
+                if (isChanged && (getLog().isDebugEnabled() || showCompilationChanges)) {
+                    getLog().info("\tNew dependency detected: " + file.toAbsolutePath());
+                }
+                return isChanged;
+            } catch (IOException ex) {
+                // we just cannot determine it, so don't do anything beside logging
+                getLog().warn("I/O error reading the lastModifiedTime: " + ex.getMessage());
             }
         }
 
         return false;
     }
 
+    private static String getFileExtension(Path path) {
```

Comment:
Don't we have already some plexus utils for that? Or the dependency have been removed?

Reply:
Plexus-utils is not included, there is one in maven-shared-utils but it's deprecated in favor of commons-io, yet commons-io is an indirect dependency, so we can make commons-io a direct dependency for that single method or simply duplicate the method here.

Reply:
I don't mind to make commons-io a direct dependency so will change this ASAP.

I'm not on a computer right now, so I will need some time to make all the requested changes.

Thanks for all the feedback! 👍